lib/shoulda/matchers/active_model/numericality_matchers/submatchers.rb

```diff
@@ -13,29 +13,46 @@ def matches?(subject)
             failing_submatchers.empty?
           end
 
+          def does_not_match?(subject)
+            @subject = subject
+            not_failing_submatchers.empty?
+          end
+
           def failure_message
-            last_failing_submatcher.failure_message
+            failing_submatcher.failure_message
           end
 
           def failure_message_when_negated
-            last_failing_submatcher.failure_message_when_negated
+            not_failing_submatcher.failure_message_when_negated
           end
 
           def add(submatcher)
             @submatchers << submatcher
           end
 
-          def last_failing_submatcher
-            failing_submatchers.last
-          end
-
           private
 
           def failing_submatchers
             @_failing_submatchers ||= @submatchers.reject do |submatcher|
               submatcher.matches?(@subject)
             end
           end
+
+          def not_failing_submatchers
+            @_not_failing_submatchers ||= @submatchers.reject do |submatcher|
+              submatcher.does_not_match?(@subject)
+            end
+          end
+
+          def failing_submatcher
+            failing_submatchers.last
+          end
+
+          def not_failing_submatcher
+            not_failing_submatchers.detect { |submatcher|
+              submatcher.instance_of?(DisallowValueMatcher)
+            } || not_failing_submatchers.first
+          end
```

Comment:
That was something that bothered me for a while when creating the first specs for the `validate_comparison_of` matcher, sometimes the failure message was very hard to understand because it was using `AllowValueMatcher` failure messages, the `DisallowValueMatcher` ones are easy to understand when dealing with matcher that didn't fail but was supposed to.

Reply:
Hmm... I'm wondering if perhaps we should leave it to the developer to add matchers in the order that would lead to a better user experience instead of trying to "fix" it for them. I can see this behavior being somewhat confusing as well. Is there another approach we could take here or what do you think?

Reply:
This is tricky because the developer can't define the order of the submatchers. After all, in this specific case, when the user defines one comparison matcher, we define three submatchers for that matcher, and that order is defined by the `shoulda-matchers` code.

Generates
```ruby
validates_comparison_of :participant_count, greater_than_or_equal_to: 2
# Generates in shoulda-matchers
# ComparisonMatcher with the following submatchers
# DisallowValueMatcher with value 1
# AllowValueMatcher with value 2
# AllowValue matcher with value 2
```

I totally get what you're saying about this behavior being confusing. Honestly, I'm fresh out of ideas here. I'm also fine reverting that.

Reply:
I reverted that change on [5906112](https://github.com/thoughtbot/shoulda-matchers/pull/1552/commits/59061129aa5913caa1536f1d3f76e8b70473f089)

Reply:
Ah sorry for not replying, I'm slammed with work currently 😵‍💫 Thank you! If this indeed continues to be confusing, I'm all for revisiting this. It'll also be easier to see the consequences of this change separately.